processor/transformer/transformer.go

```diff
@@ -296,72 +301,46 @@ func (trans *HandleT) request(ctx context.Context, url string, data []Transforme
 		return nil
 	}
 
-	// assume that the first event is representative
-
-	for {
-		s := time.Now()
-		trace.WithRegion(ctx, "request/post", func() {
-			resp, err = trans.Client.Post(url, "application/json; charset=utf-8", bytes.NewBuffer(rawJSON))
-		})
-		if err == nil {
-			// If no err returned by client.Post, reading body.
-			// If reading body fails, retrying.
-			respData, err = io.ReadAll(resp.Body)
-			resp.Body.Close()
-		}
-
-		if err != nil {
-			trans.requestTime(statsTags(data[0]), time.Since(s))
-			reqFailed = true
-			trans.logger.Errorf("JS HTTP connection error: URL: %v Error: %+v", url, err)
-			if retryCount > maxRetry {
-				panic(fmt.Errorf("JS HTTP connection error: URL: %v Error: %+v", url, err))
-			}
-			retryCount++
-			time.Sleep(retrySleep)
-			// Refresh the connection
-			continue
-		}
-		if reqFailed {
-			trans.logger.Errorf("Failed request succeeded after %v retries, URL: %v", retryCount, url)
-		}
+	var (
+		respData   []byte
+		statusCode int
+	)
 
-		// perform version compatibility check only on success
-		if resp.StatusCode == http.StatusOK {
-			transformerAPIVersion, convErr := strconv.Atoi(resp.Header.Get("apiVersion"))
-			if convErr != nil {
-				transformerAPIVersion = 0
+	// endless retry if transformer-controlplane connection is down
+	endlessBackoff := backoff.NewExponentialBackOff()
+	endlessBackoff.MaxElapsedTime = 0 // no max time -> ends only when no error
+	// endless backoff loop, only nil error or panics inside
+	_ = backoff.RetryNotify(
+		func() error {
+			respData, statusCode = trans.doPost(ctx, rawJSON, url, statsTags(data[0]))
+			if statusCode == StatusCPDown {
+				trans.cpDownGauge.Gauge(1)
+				return fmt.Errorf("control plane not reachable")
 			}
-			if types.SUPPORTED_TRANSFORMER_API_VERSION != transformerAPIVersion {
-				trans.logger.Errorf("Incompatible transformer version: Expected: %d Received: %d, URL: %v", types.SUPPORTED_TRANSFORMER_API_VERSION, transformerAPIVersion, url)
-				panic(fmt.Errorf("Incompatible transformer version: Expected: %d Received: %d, URL: %v", types.SUPPORTED_TRANSFORMER_API_VERSION, transformerAPIVersion, url))
-			}
-		}
-
-		trans.requestTime(statsTags(data[0]), time.Since(s))
-		break
-	}
+			trans.cpDownGauge.Gauge(0)
+			return nil
+		},
+		endlessBackoff,
+		func(err error, t time.Duration) {
+			trans.logger.Errorf("JS HTTP connection error: URL: %v Error: %+v", url, err)
+		})
+	// control plane back up
 
 	// Remove Assertion?
-	if !(resp.StatusCode == http.StatusOK ||
-		resp.StatusCode == http.StatusBadRequest ||
-		resp.StatusCode == http.StatusNotFound ||
-		resp.StatusCode == http.StatusRequestEntityTooLarge) {
-		trans.logger.Errorf("Transformer returned status code: %v", resp.StatusCode)
+	if !(statusCode == http.StatusOK ||
+		statusCode == http.StatusBadRequest ||
+		statusCode == http.StatusNotFound ||
+		statusCode == http.StatusRequestEntityTooLarge) {
+		trans.logger.Errorf("Transformer returned status code: %v", statusCode)
```

Comment:
It can be simplified by using a switch case:

```suggestion
        switch statusCode {
        case http.StatusOK, http.StatusBadRequest, http.StatusNotFound,http.StatusRequestEntityTooLarge:
		trans.logger.Errorf("Transformer returned status code: %v", statusCode)
```
It is a matter of preference ofc

Reply:
you are missing a default case statement where the log statement should belong though, right? :)